docs: add upgrade/release notes for 1.12.0 (#13269)
hashicorp/packer: website/content/docs/upgrade/v1_12.mdx

Packer 1.12's notable changes can be summarised by the following points:

1. Protocol changes for Packer/Plugin communication
2. Introduction of a DAG for Data Sources and Locals

**Note**: None of the changes we made should impact your templates, and they should continue working as-is.

## Protocol changes

When initialising, Packer is now able to pick between two different protocols for communication with Plugins: gob (legacy), and protobuf/msgpack.

The latter is to become the standard after this release, but in order to avoid breaking retro-compatibility, it will only be picked if the following are true:

1. All the plugin candidates for a build support the new protocol
2. Packer supports the protocol

This change should be transparent to you, as this only changes how Packer operates under the hood.
However, if you start seeing errors with Packer/Plugin communication like serialisation errors, please [report it to us](https://github.com/hashicorp/packer/issues) so we can fix it in a later version.
In the meantime however, you can set the `PACKER_FORCE_GOB` environment variable to force Packer and plugins to use the legacy serialisation format for communication.

> Example: `export PACKER_FORCE_GOB=1` will force using gob for serialisation.

Reviewer: ```suggestion
Example: `export PACKER_FORCE_GOB=1` will force using gob for serialization.
```

## Data Source and Locals DAG support

In order to allow more flexibility in how you may define data sources and locals, we introduced a graph-based approach to scheduling evaluation of those components.
This means that with this change, you can use a local variable in a data source, and vice-versa, you are not bound to a strict evaluation order.

If your templates worked with older versions of Packer, this change should be transparent to you.
However, if you are experiencing errors with your data sources or locals, typically errors due to unkown values, or other dependency problems, please [report it to us](https://github.com/hashicorp/packer/issues) so we can fix it in a later version.

> In the meantime however, the `--use-sequential-evaluation` command-line option can be used to revert back to the evaluation scheduler that Packer used until v1.12.0.

Reviewer: ```suggestion
You can use the `--use-sequential-evaluation` command-line option to revert back to the evaluation scheduler that Packer used prior to v1.12.0.
```

> Example: `packer build --use-sequential-evaluation <TEMPLATE>` will force using the sequential evaluation methods from Packer until v1.12.0.

Reviewer: ```suggestion
Example: `packer build --use-sequential-evaluation <TEMPLATE>` will force Packer to use the sequential evaluation methods.
```
